Mod Audio overhaul and fixes

Cooliokid956 opened the pull request:
THIS IS THE MOD AUDIO OVERHAUL WOOOOO

## NEW autogen feature: `OVERLOAD(name)`
allows you to combine multiple functions under a single name

# Mod audio changes:
- New function for [get/set] parity: `audio_stream_get_loop_points`
- Arguably better method for managing mod master/stream/sample volumes (sound groups, tie in very nicely with the post-established volume groups)
- Made Lua volume control functions use u8 instead of f32 (that's what they are intended to be)
- check out my new modaudio api 😁 😆
<img width="347" height="973" alt="image" src="https://github.com/user-attachments/assets/3aac0011-4aa7-46d7-bb8b-aaaf4e5b946b" />

This is my test mod that showcases some features in the new API! :) :)
[testy.zip](https://github.com/user-attachments/files/28535157/testy.zip)

Cooliokid956 committed:
Miscellaneous Audio Tweaks

- New function for get/set parity: `audio_stream_get_loop_points`
- Arguably better method for managing mod master/stream/sample volumes (sound groups)
- Made *those* functions use u8 instead of f32 (that's what they are intended to be)
- added various defines to trim some unused fat from miniaudio
- LuaTable return type for functions that needed it

Remark:
the `MA_SOUND_FLAG_STREAM` and `MA_SOUND_FLAG_DECODE` flags don't actually do anything since all audio is always read to memory before decoding (via data source), these flags only get a response in the Resource Manager, which we don't ever use here
Right now, the only true difference between samples and streams is in `MA_SOUND_FLAG_NO_PITCH` (also not really, since this flag is overturned when sample rates mismatch (very often)), as well as samples' special duplication tech

PeachyPeachSM64 commented on src/pc/lua/utils/smlua_audio_utils.c:
```diff
@@ -19,6 +19,7 @@
 #include "pc/pc_main.h"
 #include "pc/fs/fmem.h"
 
+#pragma region m64
```
`#pragma region` is a Visual Studio exclusive preprocessor code
please remove it

Cooliokid956 replied:
Ok 😢

Cooliokid956 commented on src/pc/lua/utils/smlua_audio_utils.c:
```diff
@@ -423,13 +429,30 @@ void audio_stream_set_looping(struct ModAudio* audio, bool looping) {
+    lua_rawseti(L, -2, 2);
+
+    return smlua_to_lua_table(L, -1);
+}
```
I would've done that but autogen doesn't allow it... maybe with a `RETURN(n)` macro?

Cooliokid956 replied:
Well that was quite the exercise :D 
<img width="488" height="62" alt="image" src="https://github.com/user-attachments/assets/331c47b4-a573-43ca-a2fa-194fae424e97" />
<img width="763" height="124" alt="image" src="https://github.com/user-attachments/assets/84527eb8-46bf-4ea7-bc2e-1b73e5815db6" />
<img width="778" height="218" alt="image" src="https://github.com/user-attachments/assets/68442cba-1b8d-4fc7-9839-b71c9b99784d" />
I believe that `gfx_get_from_name` and `vtx_get_from_name` can be relocated to `smlua_gfx_utils.c` now

Cooliokid956 committed:
pragma balls

Cooliokid956 committed:
Adjust for new autogen changes

idiotforge commented:
pragma balls is a good commit name

Cooliokid956 committed:
Eliminate baseVolume

Not needed

Isaac0-dev reviewed (commented):
in my opinion this is more of a "miscellaneous tweaks" pr in general. I agree with the non audio related adjustments, but they aren't related to audio. can we separate them please?

djoslin0 commented:
I agree with Isaac, split the PR into multiple. Only have one PR per enhancement. Combining stuff makes it harder to review

Cooliokid956 committed:
another autogen fix

Cooliokid956 committed:
revert unrelated change

Isaac0-dev commented:
merge conflicts!

Cooliokid956 committed:
Merge branch 'dev' into uae

Cooliokid956 committed:
Run Autogen

Cooliokid956 committed:
No more tweaking

Cooliokid956 commented:
I have decided that I will put a bit more into this

Cooliokid956 committed:
OK

Cooliokid956 committed:
Great fixes and progress

all channels work!
next I will work on replacing `bool isStream` with `u8 flags` (includes type and other things down the line probably)

Cooliokid956 committed:
No More `isStream` or `loaded`

`flags` reigns supreme

Cooliokid956 committed:
No more unction

if you were you unhide deprecated fields but we're good like this (this is also just more intelligent in general)

also hide bit fields for reasons

Cooliokid956 committed:
More progress

- brought bit fields back
- suppressed bit fields in lua structs
- i think this is good

- May have fixed a leak???Idk
It's too much

Cooliokid956 committed:
The Revolution

I believe I have given some power to samples

The beginnings of ambiguity...

Cooliokid956 committed:
A few things

real quick upload Sorry

Blockyyy committed:
the fix™️

Cooliokid956 committed:
Totally redo block'y pr

a direct access (after dereferencing the sound) to the sample rate is possible, no need to store it (look at the functions themselves!)

and a bit more progress
I will return later

Cooliokid956 committed:
MORE

might not need that At end

Cooliokid956 committed:
functions have been desegregated (mostly)

- Fixed an autogen typo I thought I'd fixed long ago
- reordered ModAudio docs
- Brand new, Copy Audio.
- reordered audio utils
- ToDo! add pan property and make a test mod

it is 5:02:17. Good night

Cooliokid956 committed:
It's just intrusive

I have decided to abandon the No Pitch flag as it presented a few bugs in my test mod. Fly High `MA_SOUND_FLAG_NO_PITCH`

Cooliokid956 committed:
.playing

Cooliokid956 committed:
More cleaning? But there's still a mess...

decoders are now uninited as well
Added panning
Added old functions to hidden builtin
Always free some things
Always assign volumes

Cooliokid956 committed:
Overload start

Cooliokid956 committed:
more bits

Cooliokid956 committed:
Add sanity checks

and type of None

Cooliokid956 committed:
detect overloaded functions

TODO: process them as such

Cooliokid956 committed:
Merge remote-tracking branch 'upstream/dev' into uae

Cooliokid956 committed:
Merge remote-tracking branch 'upstream/dev' into uae

Cooliokid956 committed:
Finalize merge

Cooliokid956 committed:
`ModAudio::reload()`

- also moved deprecated function definitions to the proper file
- also fixed sample copy playback (simple typo)

Cooliokid956 committed:
Merge remote-tracking branch 'upstream/dev' into uae

Cooliokid956 committed:
new standard for overloading

Cooliokid956 committed:
OVERLOADING

FINALLY
also fixed some autogen formatting along the way

still need that range check block... but it's not like I *need* it... Some other time

Cooliokid956 committed:
Almost final changes

- restore `djui_hud_print_text` and `djui_hud_print_text_interpolated` using new overload technology
- condense `audio_stream_load` and `audio_sample_load` using old deprecation technology
  - `audio_load(filename, type)`
- fixed off-by-one in overload param check selection
- refactored audio test example mod for new functions

Cooliokid956 committed:
add param count checks

Cooliokid956 committed:
length

- remove debug prints
- further upgrades to the audio test
- remove redundant casts

Cooliokid956 committed:
nice catch

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

Cooliokid956 committed:
kill debug prints

I genuinely did not need them
also reordered bitflags for some reason?

Cooliokid956 committed:
Rerun autogen

missed this single flag...

PeachyPeachSM64 commented on src/pc/djui/djui_hud_utils.c:
```diff
@@ -722,6 +722,16 @@ void djui_hud_print_text(const char* message, f32 x, f32 y, f32 scaleX, f32 scal
+        scale *= 0.5f;
+    }
+
+    djui_hud_print_text_internal(message, x, y, scale, scale, NULL);
```
please do `djui_hud_print_text(message, x, y, scale, scale)` instead

PeachyPeachSM64 commented on src/pc/djui/djui_hud_utils.c:
```diff
@@ -750,6 +760,32 @@ void djui_hud_print_text_interpolated(const char* message, f32 prevX, f32 prevY,
+        interp->height = font->defaultFontScale;
+    }
+
+    djui_hud_print_text_internal(message, x, y, scale, scale, interp);
```
same here
`djui_hud_print_text_interpolated(message, prevX, prevY, prevScale, prevScale, x, y, scale, scale)`

PeachyPeachSM64 commented on src/pc/lua/utils/smlua_audio_utils.c:
```diff
@@ -182,20 +182,171 @@ u8 smlua_audio_utils_allocate_sequence(void) {
+#define MA_SOUND_FLAGS (MA_SOUND_FLAG_NO_SPATIALIZATION | MA_SOUND_FLAG_NO_DEFAULT_ATTACHMENT)
 
 static ma_engine sModAudioEngine;
+static const char* sModAudioTypes[] = { "sound", "sample", "stream" };
```
`static const char *sModAudioTypes`

PeachyPeachSM64 commented on src/pc/lua/utils/smlua_audio_utils.c:
```diff
@@ -182,20 +182,171 @@ u8 smlua_audio_utils_allocate_sequence(void) {
 static struct DynamicPool *sModAudioPool;
 static bool sModAudioShuttingDown = false;
 
+static bool audio_sanity_check(struct ModAudio* audio, u8 type, const char* action) {
```
`struct ModAudio *audio`
`const char *action`

PeachyPeachSM64 commented on src/pc/lua/utils/smlua_audio_utils.c:
```diff
@@ -182,20 +182,171 @@ u8 smlua_audio_utils_allocate_sequence(void) {
+
+// Called whenever a sample copy finishes playback (called from the miniaudio thread)
+// removes the copy from its linked list, and adds it to the pending list
+static void audio_destroy_copy(struct ModAudio* copy) {
```
`struct ModAudio *copy`

PeachyPeachSM64 commented on src/pc/lua/utils/smlua_audio_utils.c:
```diff
@@ -182,20 +182,171 @@ u8 smlua_audio_utils_allocate_sequence(void) {
+    pthread_mutex_unlock(&sSoundCopyMutex);
+}
+
+static void audio_destroy_copy_callback(void* userData, UNUSED ma_sound* sound) {
```
well, pointers
I'm not filling the entire review with these
`void *userData`, NOT `void* userData`

PeachyPeachSM64 commented on src/pc/lua/utils/smlua_audio_utils.c:
```diff
@@ -182,20 +182,171 @@ u8 smlua_audio_utils_allocate_sequence(void) {
+        && copy && copy->copy
+        && copy->parent->alive
+    ) {
+        if (copy->next) { copy->next->prev = copy->prev; }
+        if (copy->prev) { copy->prev->next = copy->next; }
+        if (copy == copy->parent->copiesTail) {
+            copy->parent->copiesTail = copy->prev;
+        }
+        copy->next = NULL;
+        copy->prev = NULL;
+        copy->parent = NULL;
+    
+        // add copy to free list
+        if (sSoundCopyFreeTail) {
+            copy->prev = sSoundCopyFreeTail;
+            sSoundCopyFreeTail->next = copy;
+        }
+        sSoundCopyFreeTail = copy;
```
I hate linked lists with a passion just for this kind of non-sense
But, oh well, if it works, it works

Cooliokid956 replied:
I didn't do this

PeachyPeachSM64 commented on src/pc/lua/utils/smlua_audio_utils.c:
```diff
@@ -182,20 +182,171 @@ u8 smlua_audio_utils_allocate_sequence(void) {
         LOG_ERROR("failed to init Miniaudio: %d", result);
     }
+
+    for (u8 i = 0; i < 3; i++) {
```
ARRAY_COUNT(sModAudioChannels) instead of 3

PeachyPeachSM64 commented on src/pc/lua/utils/smlua_audio_utils.c:
```diff
@@ -393,399 +518,279 @@ struct ModAudio* audio_load_internal(const char* filename, bool isStream) {
+        f32 dX = position[0] - gCamera->pos[0];
+        f32 dY = position[1] - gCamera->pos[1];
+        f32 dZ = position[2] - gCamera->pos[2];
+        dist = sqrtf(dX * dX + dY * dY + dZ * dZ);
```
or just `dist = vec3f_dist(position, gCamera->pos)`

PeachyPeachSM64 commented on src/pc/lua/utils/smlua_audio_utils.c:
```diff
@@ -795,32 +800,28 @@ void audio_custom_shutdown(void) {
-
+            audio->copiesTail = NULL;
+            pthread_mutex_unlock(&sSoundCopyMutex);
+            
```
whitespaces

Cooliokid956 replied:
Ok

PeachyPeachSM64 commented on src/pc/lua/utils/smlua_audio_utils.c:
```diff
@@ -831,29 +832,30 @@ void audio_custom_shutdown(void) {
-
+        pthread_mutex_unlock(&sSoundCopyMutex);
+        
+        for (u8 i = 0; i < 3; i++) { ma_sound_group_uninit(&sModAudioChannels[i]); }
```
ARRAY_COUNT(sModAudioChannels)

PeachyPeachSM64 commented on src/pc/utils/miniaudio.h:
```diff
@@ -15,13 +15,15 @@ GitHub:        https://github.com/mackron/miniaudio
 extern "C" {
 #endif
 
+// OUR DEFINES
 #define MA_NO_RESOURCE_MANAGER
 #define MA_NO_GENERATION
 #define MA_NO_ENCODING
 
 // enable Vorbis decoding (provides ogg audio decoding support) for miniaudio
 #define STB_VORBIS_HEADER_ONLY
 #include "stb_vorbis.c"
+// OUR DEFINES
```
what does that mean?

Cooliokid956 replied:
It means that these are defines that WE (I) manually added to the file and must preserve when updating miniaudio
They were defined to not include unused junk in the build

Cooliokid956 committed:
peachy #1

Cooliokid956 committed:
peachy #2: conform to repo pointer convention

Cooliokid956 committed:
peachy #3: actually yeah

Cooliokid956 committed:
peachy #4: whitespaces

Cooliokid956 committed:
peachy #5: array count

Cooliokid956 committed:
Kill All Whitespaces

xLuigiGamerx reviewed (commented):
Merge conflicts

Cooliokid956 committed:
Merge remote-tracking branch 'upstream/dev' into uae

Cooliokid956 committed:
exposed lists

xLuigiGamerx reviewed (approved):
Merge resolutions

Cooliokid956 committed:
Merge branch 'dev' into uae

EmeraldLoc reviewed (approved):
Tested and works, nothing looks wrong except possibly using `pthread_mutex_lock/unlock` functions raw rather than creating a `ThreadHandle` and using the `MUTEX_LOCK/UNLOCK` macros, but in this case it looks fine because mini audio is handling the actual thread where the func is called, not coop. Api looks a good bit cleaner so I like this. I did NOT review any code related to autogen changes, I looked at the audio stuff and tested that and that only
